### ISSUE TITLE
:- syntax awkward for methods with multiple return values or no results

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#446]
The error handling support in Dafny is currently centered around the "let or fail" statement, which attempts to `Extract()` a value and immediately returns the result of `PropagateFailure()` otherwise:

```
var x :- SomeMethod(...);
```

This doesn't compose well with methods that return multiple values, however, since `Extract()` must produce a single value, so you can never write this:

```
var x, y, z :- SomeMethod(...);
```

There is also use case for methods that operate by side-effects and return no values but may fail. I'm finding that I'm defining several methods that return a `Result<()>` of some kind and calling them like this:

```
var _ :- SomeMethod(...);
```

It's a minor annoyance but it adds up, especially when trying to implement tests. I'd suggest we support the unary variant of `:-` (currently only usable in functional expressions) for statements like this too:

```
:- SomeMethod(...);
```

**Comment by seanmcl** [comment]
I think the multiple value problem will get fixed by the pattern matching work that is nearing completion.  That will allow any (arbitrarily nested) constructors, not just tuples, on the left.  

I'm not a fan of `:- sideEffectingExpression()`, mainly because I like seeing that I'm deliberately ignoring the return value.

**Comment by robin-aws** [comment]
Interesting - does that mean the pattern matching changes will allow deconstructing assignment such as this?

```
function Foo(): (int, int) {
  (1, 2)
}

var (x, y) := Foo();
```

The only reason I'm not a big fan of `var _ :- ...` is because it bulks things up when making a lot of calls like that. I'm working on a "Require" utility function for dynamic checks/testing, and multiple uses end up looking like this (ignoring the fact that you could collapse the requirements in this example):

```
var _ :- Require(x >= 0);
var _ :- Require(y >= 0);
var _ :- Require(x < y);
...
```

I agree now that `:- X` is not the best choice since in a functional context that *doesn't* ignore the return value. 

I might be able to address this with something like a `RequireAll(tests: seq<Requirement>)` function. I'm using `Requirement` there to attach a string to failures for help in debugging - otherwise I could just do `Require(A && B &&...)`.

Another way that relying on `:-` for error handling doesn't compose well with methods with no return value: it's way too easy to do this by mistake, in which case failures are completely ignored:

```
Require(x >= 0);
Require(y >= 0);
Require(x < y);
```

**Comment by seanmcl** [comment]
Yeah, I like the idea of making a method to run a bunch of side-effecting computations in a list, combining them in the obvious way.  This is called `sequence`  in Scala's Cats and Scalaz libraries, with type something like `F[G[A]] => G[F[A]]` where `F` is traversable and `G` is applicative.  

With respect to your last issue, languages like OCaml and Standard ML warn you if you discard a non-unit value.  Perhaps Dafny should do that as well.

**Comment by RustanLeino** [comment]
@robin-aws Dafny already supports

    var (x, y) := Foo();

The left-hand side can be the same kind of patterns that follow `case` in a `match` construct. For example, if you know that `MyList()` will return a list of length at least 2, then you can destruct it like this:

    var Cons(a, Cons(b, rest)) := MyList();

However, the `:-` assignment statement does not yet support patterns on the left.

**Comment by RustanLeino** [comment]
I agree with @seanmcl that explicitly showing that you're wanting to throw away the return value is good.

@robin-aws Does your `Require` have a meaning return value? If not, you can declare its return type like the `VoidOutcome` type in `Test/exceptions/VoidOutcome.dfy`. Then, you can write

    :- Require(x >= 0);

See `Test/exceptions/Exceptions1.dfy` for examples.

**Comment by robin-aws** [comment]
No it doesn't have a meaningful return value, and I independently invented something close to `VoidOutcome` after cutting this issue. Then I discovered the `Test/exceptions` tests. :)

I had tried the unary `:-` form before but it wouldn't compile for me because my result type defines an `Extract()` method, and the compiler currently rejects the RHS in that case (again to avoid accidentally ignoring a return value). That's why you need something like `VoidOutcome`, but that means you can't use `:- Require(...)` in a method that DOES return a meaningful value - you get an error like `RHS (of type VoidOutcome) not assignable to LHS (of type Result<Foo>)`

My current solution is to use `Result<()>` as the result of `Require` instead, which means no unary `:-`, but explicitly discarding the unit result value works around that.

**Comment by robin-aws** [comment]
Perhaps the unary `:-` form should allow RHS types that return the unit type?

**Comment by RustanLeino** [comment]
If the unary `:-` would be allowed when the RHS type is `()` (which I'm not necessarily advocating), then it seems that that ought to be available for generally. That is, if `x := M(...);` is legal for an `x` of type `()`, then `x := M(...);` would be allowed to be abbreviated as just `M(...);`.

There's one other built-in type with a single value, namely `bv0`. One could argue that any abbreviation available for `()` should also be available for `bv0`.

**Comment by RustanLeino** [comment]
On this subject, let me mention two other issues where the current `:-` design doesn't do so well.

### Ghost return values

As mentioned above, if you have multiple "normal" (i.e., non-failure) out-parameters, then you can put them into a tuple, like `Result<(X,Y)>`. A caller then looks like

    var pair :- M(...);
    var (x, y) := pair;

If patterns were available as left-hand sides of `:-` (which they should be, but that hasn't been implemented yet), then you could write

    var (x, y) := M(...);

But what if one of the two values is ghost? Then, you can't use a built-in tuple type, because tuples have all non-ghost arguments. Instead, you have to declare your own record type:

    datatype GhostlyPair<X,Y> = GhostlyPair(X, ghost Y)

and then use `Result<GhostlyPair<X,Y>>`.

### Constructors

Constructors in Dafny always have to construct the object. That is, there's no way for the constructor body to bag out. That may be a good language design choice. But one could imagine a design where a constructor for a class `C` instead got to have an outcome of `Result<C>`. The body of the constructor could then choose to fail, perhaps by doing something like

    return Failure("construction currently unavailable");

The caller would know that it's calling a constructor with a `Result<C>` outcome (rather than the default `C` outcome). The caller could then do:

    var r := new C(...);

where `r` has type `Result<C>`, or could use `:-`, as in

    var c :- new C(...);

where `c` has type `C`.

**Comment by samuelgruetter** [comment]
Happy to see that someone started using this feature :smiley: 

Let me just say that not everything which we intended to implement is implemented yet, my internship was over before I could finish it...
As you can see in the [design document](https://github.com/dafny-lang/rfcs/blob/250686552d3c2aeec8fc471229083a9f1e2c2047/Exceptions.md#language-level-changes), patterns and unary `:-` for statements was definitely part of the plan.

**Comment by robin-aws** [comment]
Note that PR #455 attempts to align the `:-` with standard monadic constructs, which I think gets at the heart of these issues. My issues with the error-handling semantics when there is no return value are really about the fact that `:-` can be thought of as syntactic sugar for a monadic `Bind` method (like `>>=` in Haskell) but there is no version for the common alternative with no bound variable (like `>>` in Haskell).

**Comment by davidcok** [comment]
Fixed by PR #809
